Sources/NIOWebsocket/WebsocketUpgrader.swift

```diff
@@ -0,0 +1,127 @@
+    private let shouldUpgrade: (HTTPRequestHead) -> HTTPHeaders?
+    private let upgradePipelineHandler: (Channel, HTTPRequestHead) -> EventLoopFuture<Void>
+
+    /// Create a new `WebsocketUpgrader`.
+    ///
+    /// - parameters:
+    ///     - shouldUpgrade: A callback that determines whether the websocket request should be
+    ///         upgraded. This callback is responsible for creating a `HTTPHeaders` object with
+    ///         any headers that it needs on the response *except for* the `Upgrade`, `Connection`,
+    ///         and `Sec-Websocket-Accept` headers, which this upgrader will handle. Should return
+    ///         `nil` if the upgrade should be refused.
+    ///     - upgradePipelineHandler: A function that will be called once the upgrade response is
+    ///         flushed, and that is expected to mutate the `Channel` appropriately to handle the
+    ///         websocket protocol. This only needs to add the user handlers: the
+    ///         `WebsocketFrameEncoder` and `WebsocketFrameDecoder` will have been added to the
+    ///         pipeline automatically.
+    public init(shouldUpgrade: @escaping (HTTPRequestHead) -> HTTPHeaders?,
+                upgradePipelineHandler: @escaping (Channel, HTTPRequestHead) -> EventLoopFuture<Void>) {
+        self.shouldUpgrade = shouldUpgrade
+        self.upgradePipelineHandler = upgradePipelineHandler
+    }
+
+    public func buildUpgradeResponse(upgradeRequest: HTTPRequestHead, initialResponseHeaders: HTTPHeaders) throws -> HTTPHeaders {
+        let key = try upgradeRequest.headers.getNonListHeader("Sec-WebSocket-Key").trimmingCharacters(in: .whitespaces)
+        let version = try upgradeRequest.headers.getNonListHeader("Sec-WebSocket-Version").trimmingCharacters(in: .whitespaces)
+
+        // The version must be 13.
+        guard version == "13" else {
+            throw NIOWebsocketUpgradeError.invalidUpgradeHeader
+        }
+
+        guard var extraHeaders = self.shouldUpgrade(upgradeRequest) else {
+            throw NIOWebsocketUpgradeError.unsupportedWebsocketTarget
+        }
+
+        // Cool, we're good to go! Let's do our upgrade. We do this by concatenating the magic
+        // GUID to the base64-encoded key and taking a SHA1 hash of the result.
+        let acceptValue: String
+        do {
+            var hasher = SHA1()
+            hasher.update(utf8String: key)
+            hasher.update(utf8String: magicWebSocketGUID)
+            acceptValue = String(base64Encoding: hasher.finish())
+        }
+
+        extraHeaders.replaceOrAdd(name: "Upgrade", value: "websocket")
+        extraHeaders.add(name: "Sec-WebSocket-Accept", value: acceptValue)
+        extraHeaders.replaceOrAdd(name: "Connection", value: "upgrade")
+
+        return extraHeaders
+    }
+
+    public func upgrade(ctx: ChannelHandlerContext, upgradeRequest: HTTPRequestHead) -> EventLoopFuture<Void> {
+        // TODO(cory): Why the heck doesn't this return a future? What was I doing?
```

Comment:
seems like this TODO is not valid anymore ;) You are doing the right thing now 🍺

Reply:
Well, let's say "more right", not necessarily "right". ;) Could always be doing better!